runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala

```diff
@@ -0,0 +1,314 @@
+    // a bit much overhead here because we are using the flow to represent a single response
+    val flow = Flow.fromGraph(new AkkaNettyGrpcClientGraphStage(descriptor, fqMethodName, channel, options, true))
+    Source.single(source).viaMat(flow)(Keep.right).mapMaterializedValue(_.toJava).asJava
+  }
+
+  override def updated(options: CallOptions): JavaServerStreamingRequestBuilder[I, O] =
+    new JavaServerStreamingRequestBuilder[I, O](descriptor, fqMethodName, channel, options)
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final class ScalaBidirectionalStreamingRequestBuilder[I, O](
+  descriptor: MethodDescriptor[I, O],
+  fqMethodName: String,
+  channel: Channel,
+  val options: CallOptions)
+  extends akka.grpc.scaladsl.StreamResponseRequestBuilder[Source[I, NotUsed], O]
+  with OptionsModifications[ScalaBidirectionalStreamingRequestBuilder[I, O]] {
+
+  override def invoke(request: Source[I, NotUsed]): Source[O, NotUsed] =
+    invokeWithMetadata(request).mapMaterializedValue(_ => NotUsed)
+
+  override def invokeWithMetadata(source: Source[I, NotUsed]): Source[O, Future[GrpcResponseMetadata]] = {
+    // a bit much overhead here because we are using the flow to represent a single response
+    val flow = Flow.fromGraph(new AkkaNettyGrpcClientGraphStage(descriptor, fqMethodName, channel, options, true))
+    source.viaMat(flow)(Keep.right)
+  }
+
+  override def updated(options: CallOptions): ScalaBidirectionalStreamingRequestBuilder[I, O] =
+    new ScalaBidirectionalStreamingRequestBuilder[I, O](descriptor, fqMethodName, channel, options)
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final class JavaBidirectionalStreamingRequestBuilder[I, O](
+  descriptor: MethodDescriptor[I, O],
+  fqMethodName: String,
+  channel: Channel,
+  val options: CallOptions)
+  extends akka.grpc.javadsl.StreamResponseRequestBuilder[JavaSource[I, NotUsed], O]
+  with OptionsModifications[JavaBidirectionalStreamingRequestBuilder[I, O]] {
+
+  override def invoke(request: JavaSource[I, NotUsed]): JavaSource[O, NotUsed] =
+    invokeWithMetadata(request).mapMaterializedValue(_ => NotUsed)
+
+  override def invokeWithMetadata(source: JavaSource[I, NotUsed]): JavaSource[O, CompletionStage[GrpcResponseMetadata]] = {
+    // a bit much overhead here because we are using the flow to represent a single response
+    val flow = Flow.fromGraph(new AkkaNettyGrpcClientGraphStage(descriptor, fqMethodName, channel, options, true))
+    source.asScala.viaMat(flow)(Keep.right).mapMaterializedValue(_.toJava).asJava
+  }
+
+  override def updated(options: CallOptions): JavaBidirectionalStreamingRequestBuilder[I, O] =
+    new JavaBidirectionalStreamingRequestBuilder[I, O](descriptor, fqMethodName, channel, options)
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+trait OptionsModifications[T <: OptionsModifications[T]] {
+  def options: CallOptions
+  def updated(options: CallOptions): T
+
+  def addMetadata(key: String, value: String): T = {
```

Comment:
Since later the `GrpcResponseMetadata` contains headers, perhaps `addHeader` would make more sense here?

Reply:
Yeap, I renamed when changing it into metadata. I realize I renamed it wrong though, changed it to `withHeader` as I thought it replaces, but I realize now that it should be additive and not replace, so `add` would be better.